### PR TITLE
 udevd: track idle workers in queue 

### PR DESCRIPTION
**Pull request by LadyNamedLaura**
This way we don't have to loop over our complete hashmap to find an idle worker.

due to practical reasons i made this on top of #4600, but i kept it as a separate PR since it is less trivial than #4600

### CODE REVIEW

File: src/udev/udevd.c [hunk 1]
```diff
@@ -185,6 +187,9 @@ static void worker_free(struct worker *worker) {
 
         assert(worker->manager);
 
+        if (worker->state == WORKER_IDLE) {
```
Comment (martinpitt) [inline src/udev/udevd.c:190]: No { } for single-line bodies please.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -545,25 +555,23 @@ static void worker_spawn(Manager *manager, struct event *event) {
...
-
-        HASHMAP_FOREACH(worker, manager->workers, i) {
+retry:
+        if (!LIST_IS_EMPTY(manager->idle_workers)) {
+                struct worker *worker = manager->idle_workers;
                 ssize_t count;
 
-                if (worker->state != WORKER_IDLE)
-                        continue;
+                assert(worker->state == WORKER_IDLE);
```
Comment (martinpitt) [inline src/udev/udevd.c:565]: Can this be handled more gracefully?

What's the rationale for this change, does it make udev measurably faster? My fear is that this double bookkeeping makes things more brittle, in particular forgetting to update the list in all possible places. Such assertions would then make the main udevd crash, and bad things happen.

Comment (LadyNamedLaura) [reply src/udev/udevd.c:565]: @martinpitt To be honest I haven't tested performance yet (have to set up some hw for it). If i get too it, I'll whip up some A/B testing tomorrow.
Yeah, the assert might be a bit over the top, i could just go with an if, an error log, and goto retry.

I'll test performance first, since that was pretty much the whole reason for it.

**Comment by poettering** [comment]
So this stalled, let's close this here. Please file a new PR if you intend to resubmit this.
